R/slide.R

```diff
@@ -180,3 +194,62 @@ epi_slide_one_grp = function(.data_group, index_fun, f, ..., before_num,
                            .complete = complete)
   return(mutate(.data_group, !!new_col_name := slide_values))
 }
+
+#' @importFrom rlang .data !! !!! :=
+#' @importFrom pipeR %>>%
+#' @export
+epi_slide.epi_archive = function(x, f, ..., n = 14,
+                                 align = c("right", "center", "left"), 
+                                 before, complete = FALSE,
+                                 new_col_name = "slide_value", 
+                                 new_col_type = c("dbl", "int", "lgl", "chr",
+                                                  "list"), 
+                                 time_step,
+                                 issue_step,
+                                 issue_to_max_time_value = identity,
+                                 issue_range = range(x$issues_with_updates(),
+                                                     x$max_issue())) {
+  ## TODO test this.
+
+  ## Which map function?
+  new_col_type = match.arg(new_col_type)
+  map_zzz = switch(new_col_type,
+                           "dbl" = purrr::map_dbl,
+                           "int" = purrr::map_int,
+                           "lgl" = purrr::map_lgl,
+                           "chr" = purrr::map_chr,
+                           "list" = purrr::map)
+
+  ## It'd be natural to write
+  ## ```
+  ## seq(issue_range[[1L]], issue_range[[2L]], issue_step(1))
+  ## ```
+  ## but this produces an error, as do `seq` approaches directly on `Period` objects.
+  n.issues = diff(issue_range)
+  issues = issue_range[[1L]] + issue_step(0:as.integer(lubridate::as.period(diff(issue_range))/issue_step(1L)))
+
+  ## XXX use x$naming_info()$issue.colname?
+
+  ## TODO detect skipped empty data, e.g., when trying to produce results
+  ## expecting 0 initial report latency; warn, error, or produce results
+
+  tibble::tibble(issue := issues) %>>%
+    tibble::add_column(
+              results.for.issue = . %>>%
+                dplyr::pull(issue) %>>%
+                purrr::map(function(issue) {
+                  max_time_value = issue_to_max_time_value(issue)
+                  min_time_value = max_time_value - time_step(n-1L)
+                  x$epi_df_as_of(issue) %>>%
+                    ## XXX we may want to `complete` the `time_value`s here
+                    dplyr::filter(min_time_value <= time_value & time_value <= max_time_value) %>>%
+                    dplyr::group_by(geo_value, !!!x$naming_info()$other.key.colnames) %>>%
```

Comment:
`epi_archive` is lacking a `group_by` + `epi_slide` combination to match `epi_df`.  It was written assuming that `epi_slide` would have always been called on things grouped by the key and is written to do that implicitly, but that's not actually the case.  Maybe it'd be as simple as implementing a `group_by` method to check and store group column names and then using them here, not sure.

Reply:
Thanks! My new implementation of `slide()` allows explicit specification of grouping variables, and defaults to key variables in the `epi_archive`.